File: website/static/js/comment.js
let productId = document.querySelector(".productId");
let commentDataSection = document.querySelector("#commentDataSection");
document.querySelector("#postCommentData").addEventListener("click", (e) => {
  e.preventDefault();

  let fileComment = document.querySelector("#fileComment");
  let headline = document.querySelector("#headline");
  let review = document.querySelector("#review");
  let rating = document.querySelector("#rating");

  let newComment = new FormData();
  newComment.append("fileComment", fileComment.files[0]);
  newComment.append("headline", headline.value);
  newComment.append("review", review.value);
  newComment.append("rating", rating.value);
  newComment.append("product_id", productId.id);
  fetch(`/comment?productId=${productId.id}`, {
    method: "POST",
    body: newComment,
  })
    .then((res) => res.json())
    .then((res) => {
      if (res["authenticated"] === false) {
        location.replace("/login");
      }
      commentHtml(res["results"]);
    });

  document.querySelector("#commentSection").reset();
});

const productCommentData = () => {
  fetch(`/product-comment?product=${productId.id}`)
    .then((res) => res.json())
    .then((res) => {
      let comments = res["results"];
      if (comments == true) {
        comments.forEach((comment) => {
          commentHtml(comment);
        });
      } else {
        comments.forEach((comment) => {
          commentHtml(comment);
        });
      }
    });
};

productCommentData();
// console.log(data.name[0]);
let commentOutput = "";
const commentHtml = (data) => {
  commentOutput += `
        <div class="row border-bottom py-4 commentCard${data.id}" id="${
    data.id
  }">
                            
        <div class="col-lg-1 avatar rounded-circle bg-warning ">
        ${data.name[0]}
        </div>
        <div class="col comment">
        <h5 class="mb-1">
            ${data.name}
        </h5>
        <p class="small date${data.id}"> <span class="text-muted ">${
    data.date
  }</span>
              </p>
              <h6 class="rating${data.id}"> ${
    data.rating
  } <span class="heading${data.id}">${data.heading}</span></h6>
    
        <p class="review${data.id}">${data.review}</p>
        ${
          data.image_url
            ? `
            <div class="img d-flex">
                <div class="m-2 image_data${data.id}">
                <img src="/static/img/${data.image_url}" width="100">
                </div>
            </div>
            `
            : `
            <div class="img d-flex">
                <div class="m-2 image_data${data.id}">
                </div>
            </div>
            `
        }

        <div class="update pt-3 d-flex justify-content-between">
            <div class="likeBtn">
            ${
              data.like.length == 0
                ? `
                <span id="like-count${data.id}" class="like-count">${data.like.length}</span>
                <a class="like-btn text-dark like" id="${data.id}">
                    <i onclick="likeButton(${data.id})" id="likeId${data.id}" class="bi bi-hand-thumbs-up likeComment"></i>
                </a>
                `
                : `
                <span id="like-count${data.id}" class="like-count text-success">${data.like.length}</span>   
                <a class="like-btn text-dark like" id="${data.id}">
                    <i onclick="likeButton(${data.id})" id="likeId${data.id}" class="bi bi-hand-thumbs-up-fill likeComment text-success"></i>
                </a>
                `
            }
            
            </div>
            ${
              data.user_in === data.user_id
                ? `
            <div class="nav-item dropdown me-4">
            <a class="nav-link" role="button" data-bs-toggle="dropdown" aria-expanded="false">
                <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" class="bi bi-three-dots" viewBox="0 0 16 16">
                <path d="M3 9.5a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"/>
                </svg>
            </a>
              <ul class="dropdown-menu">
                    <li>
                      <a class="dropdown-item myProdEdit updateComment" id=${data.id} href="#commentSection">
                      <i id="${data.id}" class="bi bi-pencil-square myProdEdit updateComment" ></i>
                        Edit
                      </a>
                    </li>
                    <li>
                      <a class="dropdown-item myProdDelete deleteComment" id=${data.id}  data-bs-toggle="modal" data-bs-target="#deleteComment">
                      <i id="${data.id}" class="bi bi-trash3 myProdDelete deleteComment"></i>
                        Delete
                      </a>
                    </li>
                  </ul>
                  
              
            </div> `
                : ""
            }
        </div>
        </div>
        
    </div>
    `;
  commentDataSection.innerHTML = commentOutput;
};

// LIKE FUNCTION
let likeButton = (id) => {
  let likeBtnSvg = document.querySelector(`#likeId${id}`);
  let likeCount = document.querySelector(`#like-count${id}`);
  fetch(`/like-comment?like=${id}`, { method: "POST" })
    .then((res) => res.json())
    .then((res) => {
      if (res["is_authenticated"] == false) {
        location.replace("/login");
      }
      likeCount.innerHTML = res["results"].like;
      if (res["results"].liked === true) {
        likeBtnSvg.className =
          "bi bi-hand-thumbs-up-fill likeComment text-success";
      } else {
        likeBtnSvg.className = "bi bi-hand-thumbs-up likeComment text-dark";
      }
    });
};

// DELETE FUNCTION

let newDeleteId;
let editItemId;
commentDataSection.addEventListener("click", (e) => {
  let deleteId = e.target.classList.contains("deleteComment");
  let editId = e.target.classList.contains("updateComment");

  if (editId) {
    document.querySelector("#postCommentData").style.display = "none";
    document.querySelector("#editCommentData").style.display = "block";
    let fileComment = document.querySelector("#fileComment");
    let headline = document.querySelector("#headline");
    let review = document.querySelector("#review");
    let rating = document.querySelector("#rating");
    let editComment = new FormData();
    editComment.append("fileComment", fileComment.files[0]);
    editComment.append("headline", headline.value);
    editComment.append("review", review.value);
    editComment.append("rating", rating.value);
    fetch(`/edit-comment?edit=${e.target.id}`)
      .then((res) => res.json())
      .then((res) => {
        // this then is for fetching the the comment
        let result = res["results"];
        headline.value = result.heading;
        review.value = result.review;
        rating.value = result.rating;
      });

    // itemId
    editItemId = e.target.id;
  }
  if (deleteId) {
    newDeleteId = e.target.id;
  }
});

document.querySelector("#deleteComment").addEventListener("click", () => {
  fetch(`/delete-comment?comment=${newDeleteId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ product_id: productId.id }),
  })
    .then((res) => res.json())
    .then(() => {
      let commentCardDelete = document.querySelector(
        `.commentCard${newDeleteId}`
      );

      commentDataSection.removeChild(commentCardDelete);
    });
});

// this then is for updating the comment
document.querySelector("#editCommentData").addEventListener("click", (e) => {
  e.preventDefault();

  let fileComment = document.querySelector("#fileComment");
  let headline = document.querySelector("#headline");
  let review = document.querySelector("#review");
  let rating = document.querySelector("#rating");
  let editComment = new FormData();
  editComment.append("fileComment", fileComment.files[0]);
  editComment.append("headline", headline.value);
  editComment.append("review", review.value);
  editComment.append("rating", rating.value);
  editComment.append("product_id", productId.id);
  e.preventDefault();
  fetch(`/submit-edit-comment?edit=${editItemId}`, {
    method: "POST",
    body: editComment,
  })
    .then((res) => res.json())
    .then((res) => {
      let editCommentData = res["results"];
      // change html data
      let date = document.querySelector(`.date${editItemId}`);
      let rating = document.querySelector(`.rating${editItemId}`);
      let review = document.querySelector(`.review${editItemId}`);
      let image = document.querySelector(`.image_data${editItemId}`);
      date.innerHTML = `
      <span class="text-muted ">${editCommentData.date}</span>
      `;
      review.innerHTML = editCommentData.review;
      rating.innerHTML = `
      ${editCommentData.rating} <span class="heading${editCommentData.editItemId}">${editCommentData.heading}</span>
      `;
      if (editCommentData.image_url) {
        image.innerHTML = `
        <img src="/static/img/${editCommentData.image_url}" width="100">
        `;
      }
      document.querySelector("#postCommentData").style.display = "block";
      document.querySelector("#editCommentData").style.display = "none";
      document.querySelector("#commentSection").reset();
    });
});

File: website/comment.py
from flask import Blueprint, request
from flask_login import current_user, login_required
from .models import Comment, Product, Like, Rating
from . import db
from os.path import realpath
from datetime import datetime

comment = Blueprint("comment", __name__)


def commentData(data):
    return {
        "user_id": 0 if not current_user.is_authenticated else current_user.id,
        "user_in": data.comment_id,
        "id": data.id,
        "name": data.user_comment.name,
        "image_url": data.file,
        "heading": data.heading,
        "rating": data.rating,
        "review": data.review,
        "date": data.date,
        "like": [{"author": comment.author} for comment in data.like],
        "liked": current_user.id in map(lambda x: x.author, data.like) if current_user.is_authenticated else ""
    }


@comment.route("/product-comment")
def product_comment():
    id = request.args.get("product")
    product = Product.query.get(id).comment
    return {"results": [commentData(comment) for comment in product]}


@comment.route("/comment", methods=["GET", "POST"])
def post_comment():

    if request.method == "POST":
        id = int(request.args.get("productId"))
        product = Product.query.get(id)
        ratings = product.user_rating
        user_authenticated = current_user.is_authenticated

        if not user_authenticated:
            return {"authenticated": user_authenticated}
        else:

            file = request.files.get("fileComment")
            headline = request.form.get("headline")
            review = request.form.get("review")
            rating = request.form.get("rating")
            percentage_product = 0
            if rating == "⭐":
                percentage_product = 20
            elif rating == "⭐⭐":
                percentage_product = 40
            elif rating == "⭐⭐⭐":
                percentage_product = 60
            elif rating == "⭐⭐⭐⭐":
                percentage_product = 80
            elif rating == "⭐⭐⭐⭐⭐":
                percentage_product = 100

            try:
                file.save(realpath(f"website/static/img/{file.filename}"))
                new_comment = Comment(
                    file=file.filename,
                    heading=headline,
                    review=review,
                    rating=rating,
                    date=datetime.now().strftime("%d %B %Y"),
                    comment_id=current_user.id,
                    user_product_comment=product,
                    user_comment=current_user
                )
            except AttributeError:
                new_comment = Comment(
                    heading=headline,
                    review=review,
                    rating=rating,
                    date=datetime.now().strftime("%d %B %Y"),
                    comment_id=current_user.id,
                    user_product_comment=product,
                    user_comment=current_user
                )

            db.session.add(new_comment)
            db.session.commit()

            product_rating = Rating(
                rating=percentage_product,
                comment_id=new_comment.id,
                user_product_rating=product,
                user_rating=current_user
            )
            db.session.add(product_rating)
            db.session.commit()
            return {"message": "Success", "results": commentData(new_comment)}
    return {"message": "Success"}


@comment.route("/like-comment", methods=["GET", "POST"])
def like_comment():
    id = request.args.get("like")

    if current_user.is_authenticated:
        comment = Comment.query.filter_by(id=id).first()
        likeComment = Like.query.filter_by(
            author=current_user.id, comment_id=id).first()
        if not comment:
            return {"message": "Comment does not exists."}
        elif likeComment:
            db.session.delete(likeComment)
            db.session.commit()
            return {"results": {"like": len(comment.like), "liked": current_user.id in map(lambda x: x.author, comment.like)}}
        else:
            like = Like(
                author=current_user.id,
                comment_id=id)
            db.session.add(like)
            db.session.commit()
            return {"results": {"like": len(comment.like), "liked": current_user.id in map(lambda x: x.author, comment.like)}}
    else:
        return {"is_authenticated": current_user.is_authenticated}


# edit comment
@comment.route("/edit-comment", methods=["GET", "POST"])
def edit_comment():
    id = int(request.args.get("edit"))
    comment = Comment.query.get(id)

    return {"message": "Data", "results": commentData(comment)}


@comment.route("/submit-edit-comment", methods=["GET", "POST"])
def edit_comment_data():
    id = request.args.get("edit")
    comment = Comment.query.get(int(id))
    file = request.files.get("fileComment")
    headline = request.form.get("headline")
    review = request.form.get("review")
    rating = request.form.get("rating")
    product_id = request.form.get("product_id")
    percentage_product = 0

    if rating == "⭐":
        percentage_product = 20
    elif rating == "⭐⭐":
        percentage_product = 40
    elif rating == "⭐⭐⭐":
        percentage_product = 60
    elif rating == "⭐⭐⭐⭐":
        percentage_product = 80
    elif rating == "⭐⭐⭐⭐⭐":
        percentage_product = 10

    try:
        file.save(realpath(f"website/static/img/{file.filename}"))
        comment.file = file.filename
        comment.heading = headline
        comment.review = review
        comment.rating = rating
    except AttributeError:
        comment.heading = headline
        comment.review = review
        comment.rating = rating

    # also update rating status
    update_rating = Rating.query.filter_by(comment_id=id).first()
    update_rating.rating = percentage_product

    db.session.commit()
    return {"results": commentData(comment)}

# delete comment


@comment.route("/delete-comment", methods=["GET", "POST"])
def delete_comment():
    id = int(request.args.get("comment"))
    comment_delete = Comment.query.get(id)
    like_delete = Like.query.filter_by(comment_id=id).all()
    rating_num = Rating.query.filter_by(
        comment_id=comment_delete.id).first()

    for like in like_delete:
        db.session.delete(like)
    db.session.delete(comment_delete)
    db.session.delete(rating_num)
    db.session.commit()

    return {"message": "Deleted"}
